cpp/src/parquet/column_reader_test.cc

```diff
@@ -572,5 +576,615 @@ TEST_F(TestPrimitiveReader, TestNonDictionaryEncodedPagesWithExposeEncoding) {
+  NodePtr type = schema::Int32("b", Repetition::REPEATED);
+  const ColumnDescriptor descr(type, level_info.def_level, level_info.rep_level);
+
+  // Records look like {null, [20, 20, 20], null, [30, 30], [40]}
+  std::vector<std::shared_ptr<Page>> pages;
+  std::vector<int32_t> values = {20, 20, 20, 30, 30, 40};
+  std::vector<int16_t> def_levels = {0, 1, 1, 1, 0, 1, 1, 1};
+  std::vector<int16_t> rep_levels = {0, 0, 1, 1, 0, 0, 1, 0};
+
+  std::unique_ptr<PageReader> pager;
+  std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+      &descr, values, /*num_values=*/static_cast<int>(values.size()), Encoding::PLAIN,
+      /*indices=*/{},
+      /*indices_size=*/0, def_levels, level_info.def_level, rep_levels,
+      level_info.rep_level);
+  pages.push_back(std::move(page));
+  pager.reset(new test::MockPageReader(pages));
+
+  std::shared_ptr<internal::RecordReader> record_reader =
+      internal::RecordReader::Make(&descr, level_info);
+  record_reader->SetPageReader(std::move(pager));
+
+  {
+    // This should skip the first null record.
+    int64_t records_skipped = record_reader->SkipRecords(/*num_records=*/1);
+
+    ASSERT_EQ(records_skipped, 1);
+    ASSERT_EQ(record_reader->values_written(), 0);
+    ASSERT_EQ(record_reader->null_count(), 0);
+    // For repeated fields, we need to read the levels to find the record
+    // boundaries and skip. So some levels are read, however, the skipped
+    // level should not be there after the skip. That's why levels_position()
+    // is 0.
+    ASSERT_EQ(record_reader->levels_written(), 7);
+    ASSERT_EQ(record_reader->levels_position(), 0);
+  }
+
+  {
+    // Read [20, 20, 20]
+    int64_t records_read = record_reader->ReadRecords(/*num_records=*/1);
+
+    ASSERT_EQ(records_read, 1);
+    ASSERT_EQ(record_reader->values_written(), 3);
+    ASSERT_EQ(record_reader->null_count(), 0);
+    ASSERT_EQ(record_reader->levels_written(), 7);
+    ASSERT_EQ(record_reader->levels_position(), 3);
+
+    const auto read_values = reinterpret_cast<const int32_t*>(record_reader->values());
+    std::vector<int32_t> read_vals(read_values,
+                                   read_values + record_reader->values_written());
+    std::vector<int16_t> read_defs(
+        record_reader->def_levels(),
+        record_reader->def_levels() + record_reader->levels_position());
+    std::vector<int16_t> read_reps(
+        record_reader->rep_levels(),
+        record_reader->rep_levels() + record_reader->levels_position());
+
+    ASSERT_THAT(read_vals, ElementsAre(20, 20, 20));
+    ASSERT_THAT(read_defs, ElementsAre(1, 1, 1));
+    ASSERT_THAT(read_reps, ElementsAre(0, 1, 1));
+  }
+
+  {
+    // Skip the null record and also skip [30, 30]
+    int64_t records_skipped = record_reader->SkipRecords(/*num_records=*/2);
+
+    ASSERT_EQ(records_skipped, 2);
+    ASSERT_EQ(record_reader->values_written(), 3);
+    ASSERT_EQ(record_reader->null_count(), 0);
+    // We remove the skipped levels from the buffer.
```

Comment:
Hmm, why did that happen? Is it a heuristic to compact the buffer?

Reply:
ping @fatemehp

Reply:
This is for correctness. The callers do not expect to see the skipped levels in the buffer after calling SkipRecords.